File: app/src/main/java/com/pr0gramm/app/feed/FeedService.java
package com.pr0gramm.app.feed;

import com.google.common.base.Optional;
import com.google.common.base.Strings;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.api.categories.ExtraCategoryApi;
import com.pr0gramm.app.api.categories.ExtraCategoryApiProvider;
import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.api.pr0gramm.response.Feed;
import com.pr0gramm.app.api.pr0gramm.response.Post;
import com.pr0gramm.app.services.Track;

import org.immutables.value.Value;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;

/**
 * Performs the actual request to get the items for a feed.
 */
@Singleton
public class FeedService {
    private static final Logger logger = LoggerFactory.getLogger(FeedService.class);

    private final Api mainApi;
    private final ExtraCategoryApi categoryApi;
    private final Settings settings;

    @Inject
    public FeedService(Api mainApi, ExtraCategoryApiProvider categoryApi, Settings settings) {
        this.mainApi = mainApi;
        this.categoryApi = categoryApi.get();
        this.settings = settings;
    }

    public Observable<Feed> getFeedItems(FeedQuery query) {
        FeedFilter feedFilter = query.feedFilter();
        Track.requestFeed(feedFilter.getFeedType());

        // filter by feed-type
        Integer promoted = (feedFilter.getFeedType() == FeedType.PROMOTED) ? 1 : null;
        Integer following = (feedFilter.getFeedType() == FeedType.PREMIUM) ? 1 : null;

        int flags = ContentType.combine(query.contentTypes());
        String tags = feedFilter.getTags().orNull();
        String user = feedFilter.getUsername().orNull();

        // FIXME this is quite hacky right now.
        String likes = feedFilter.getLikes().orNull();
        Boolean self = Strings.isNullOrEmpty(likes) ? null : true;

        switch (query.feedFilter().getFeedType()) {
            case RANDOM:
                return categoryApi.random(tags, flags);

            case BESTOF:
                int benisScore = settings.bestOfBenisThreshold();
                return categoryApi.bestof(tags, user, flags, query.older().orNull(), benisScore);

            case CONTROVERSIAL:
                return categoryApi.controversial(flags, query.older().orNull());

            default:
                return mainApi.itemsGet(promoted, following,
                        query.older().orNull(), query.newer().orNull(), query.around().orNull(),
                        flags, tags, likes, self, user);
        }
    }

    public Observable<Post> loadPostDetails(long id) {
        return mainApi.info(id);
    }

    @Value.Immutable
    public interface FeedQuery {
        FeedFilter feedFilter();

        Set<ContentType> contentTypes();

        Optional<Long> newer();

        Optional<Long> older();

        Optional<Long> around();
    }
}

File: app/src/main/java/com/pr0gramm/app/feed/FeedService.kt
package com.pr0gramm.app.feed

import com.pr0gramm.app.Instant
import com.pr0gramm.app.Logger
import com.pr0gramm.app.Stats
import com.pr0gramm.app.TimeFactory
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.db.FeedItemInfoQueries
import com.pr0gramm.app.services.UserService
import com.pr0gramm.app.time
import com.pr0gramm.app.ui.base.AsyncScope
import com.pr0gramm.app.ui.base.launchIgnoreErrors
import com.pr0gramm.app.util.equalsIgnoreCase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.runInterruptible
import java.util.Locale

/**
 * Performs the actual request to get the items for a feed.
 */

interface FeedService {
    suspend fun load(query: FeedQuery): Api.Feed

    suspend fun post(id: Long, bust: Boolean = false): Api.Post

    suspend fun item(itemId: Long): Api.Feed.Item {
        val query = FeedQuery(
            filter = FeedFilter().withFeedType(FeedType.NEW),
            contentTypes = ContentType.AllSet,
            around = itemId,
        )

        return load(query).items.single { item -> item.id == itemId }
    }

    /**
     * Streams feed items - giving one page after the next until
     * the end of the stream.
     */
    fun stream(startQuery: FeedQuery): Flow<Api.Feed>

    data class FeedQuery(
        val filter: FeedFilter, val contentTypes: Set<ContentType>,
        val newer: Long? = null, val older: Long? = null, val around: Long? = null
    )

}

class FeedServiceImpl(
    private val api: Api,
    private val userService: UserService,
    private val itemQueries: FeedItemInfoQueries
) : FeedService {
    private val logger = Logger("FeedService")

    override suspend fun load(query: FeedService.FeedQuery): Api.Feed {
        val feedFilter = query.filter

        // filter by feed-type
        val promoted = if (feedFilter.feedType === FeedType.PROMOTED) 1 else null
        val following = if (feedFilter.feedType === FeedType.STALK) 1 else null

        val flags = ContentType.combine(query.contentTypes)

        val feedType = feedFilter.feedType

        // statistics
        Stats().incrementCounter(
            "feed.loaded",
            "type:" + feedType.name.lowercase(Locale.ROOT)
        )

        val tags = feedFilter.tags?.replaceFirst("^\\s*\\?\\s*".toRegex(), "!")

        return when (feedType) {
            FeedType.RANDOM -> {
                // convert to a query including the x:random flag.
                val bust = Instant.now().millis / 1000L
                val tagsQuery = Tags.joinAnd("!-(x:random | x:$bust)", feedFilter.tags)

                // and load it directly on 'new'
                val feed = load(
                    query.copy(
                        newer = null, older = null, around = null,
                        filter = feedFilter.withFeedType(FeedType.NEW).basicWithTags(tagsQuery)
                    )
                )

                // then shuffle the result
                feed.copy(_items = feed._items?.shuffled())
            }

            FeedType.BESTOF -> {
                // add add s:1000 tag to the query.
                // and add s:700 to nsfw posts.

                val tagsQuery = Tags.joinOr(
                    Tags.joinAnd("s:2000", feedFilter.tags),
                    Tags.joinAnd("s:700 f:nsfw", feedFilter.tags),
                )

                load(query.copy(filter = feedFilter.withFeedType(FeedType.NEW).basicWithTags(tagsQuery)))
            }

            FeedType.CONTROVERSIAL -> {
                // just add the f:controversial flag to the query
                val tagsQuery = Tags.joinAnd("!f:controversial", feedFilter.tags)
                load(query.copy(filter = feedFilter.withFeedType(FeedType.NEW).basicWithTags(tagsQuery)))
            }

            else -> {
                val collection = feedFilter.collection
                val user = feedFilter.username

                val self = userService.loginState
                    .let { loginState ->
                        // we have a user and it is the same as in the query.
                        loginState.name != null && loginState.name.equalsIgnoreCase(user)
                    }
                    .takeIf { self -> self }

                // do the normal query as is.
                val result = api.itemsGet(
                    promoted, following,
                    query.older, query.newer, query.around,
                    flags, tags, collection, self, user
                )

                result.also {
                    AsyncScope.launchIgnoreErrors {
                        logger.time("Cache items from result") {
                            cacheItems(result)
                        }
                    }
                }
            }
        }
    }

    private suspend fun cacheItems(feed: Api.Feed) {
        runInterruptible(Dispatchers.IO) {
            itemQueries.transaction {
                for (item in feed.items) {
                    itemQueries.cache(
                        id = item.id,
                        promotedId = item.promoted,
                        userId = item.userId,
                        image = item.image,
                        thumbnail = item.thumb,
                        fullsize = item.fullsize,
                        user = item.user,
                        up = item.up,
                        down = item.down,
                        mark = item.mark,
                        flags = item.flags,
                        width = item.width,
                        height = item.height,
                        created = item.created.epochSeconds,
                        audio = item.audio,
                        deleted = item.deleted,
                    )

                    for (variant in item.variants) {
                        itemQueries.cacheVariant(
                            itemId = item.id,
                            name = variant.name,
                            path = variant.path,
                        )
                    }

                    for (subtitle in item.subtitles) {
                        itemQueries.cacheSubtitle(
                            itemId = item.id,
                            language = subtitle.language,
                            path = subtitle.path,
                        )
                    }
                }
            }
        }
    }

    override suspend fun post(id: Long, bust: Boolean): Api.Post {
        val buster = if (bust) TimeFactory.currentTimeMillis() else null
        return api.info(id, bust = buster)
    }

    override fun stream(startQuery: FeedService.FeedQuery): Flow<Api.Feed> {
        // move from low to higher numbers if newer is set.
        val upwards = startQuery.newer != null

        return flow {
            var query: FeedService.FeedQuery? = startQuery

            while (true) {
                val currentQuery = query ?: break
                val feed = load(currentQuery)
                emit(feed)

                // get the previous (or next) page from the current set of items.
                query = when {
                    upwards && !feed.isAtStart -> feed.items.maxByOrNull { it.id }
                        ?.let { currentQuery.copy(newer = it.id) }

                    !upwards && !feed.isAtEnd -> feed.items.minByOrNull { it.id }
                        ?.let { currentQuery.copy(older = it.id) }

                    else -> null
                }
            }
        }
    }
}
